applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp

```diff
@@ -680,6 +692,118 @@ double ComputeUpwindedDensity(
+    //           by Brian Nishida (1996), Section A.2.5
+
+    // read free stream values
+    const double free_stream_density = rCurrentProcessInfo[FREE_STREAM_DENSITY];
+    const double free_stream_mach = rCurrentProcessInfo[FREE_STREAM_MACH];
+    const double heat_capacity_ratio = rCurrentProcessInfo[HEAT_CAPACITY_RATIO];
+    const BoundedVector<double, 3> free_stream_velocity = rCurrentProcessInfo[FREE_STREAM_VELOCITY];
+
+    KRATOS_ERROR_IF(free_stream_mach < std::numeric_limits<double>::epsilon())
+        << "ComputeDensityDerivativeWRTVelocitySquared: free stream mach number must be larger than zero." << std::endl;
+
+    // ratio of speed of sound to free stream speed of sound
+    const double speed_of_sound_ratio = ComputeSquaredSpeedofSoundFactor<Dim, NumNodes>(localVelocitySquared, rCurrentProcessInfo);
+
+    const double free_stream_values_const = -0.5 * free_stream_density * std::pow(free_stream_mach,2.0) / inner_prod(free_stream_velocity, free_stream_velocity);
+    const double speed_of_sound_power = (2.0 - heat_capacity_ratio) / (heat_capacity_ratio - 1.0);
+
+    KRATOS_ERROR_IF((heat_capacity_ratio - 1.0) < std::numeric_limits<double>::epsilon())
+        << "ComputeDensityDerivativeWRTVelocitySquared: heat capacity ratio must not be 1.0." << std::endl;
+
+    return free_stream_values_const * std::pow(speed_of_sound_ratio, speed_of_sound_power);
+}
+
+template <int Dim, int NumNodes>
+double ComputeUpwindedDensityDerivativeWRTVelocitySquaredSupersonicAccelerating(
+    const array_1d<double, Dim>& rCurrentVelocity, 
+    const array_1d<double, Dim>& rUpwindVelocity, 
+    const ProcessInfo& rCurrentProcessInfo)
+{
+    // Following Fully Simulataneous Coupling of the Full Potential Equation
+    //           and the Integral Boundary Layer Equations in Three Dimensions
+    //           by Brian Nishida (1996), Section A.2.6
+    const double Drho_Dq2 = ComputeDensityDerivativeWRTVelocitySquared<Dim, NumNodes>(inner_prod(rCurrentVelocity, rCurrentVelocity), rCurrentProcessInfo);
+
+    const double upwind_factor = SelectMaxUpwindFactor<Dim, NumNodes>(rCurrentVelocity, rUpwindVelocity, rCurrentProcessInfo);
+
+    const double upwind_factor_derivative = ComputeUpwindFactorDerivativeWRTVelocitySquared<Dim, NumNodes>(rCurrentVelocity, rCurrentProcessInfo);
+
+    const double current_mach_sq = ComputeLocalMachNumberSquared<Dim, NumNodes>(rCurrentVelocity, rCurrentProcessInfo);
+    const double upwind_mach_sq = ComputeLocalMachNumberSquared<Dim, NumNodes>(rUpwindVelocity, rCurrentProcessInfo);
+
+    const double current_density = ComputeDensity<Dim, NumNodes>(current_mach_sq, rCurrentProcessInfo);
+    const double upwind_density = ComputeDensity<Dim, NumNodes>(upwind_mach_sq, rCurrentProcessInfo);
+
+    const double bracket_term = upwind_factor_derivative * (current_density - upwind_density) + upwind_factor * Drho_Dq2;
+
+    return Drho_Dq2 - bracket_term;
```

Comment:
I would simplify these two lines (also in the formula that you provide in the table taking Drho Dq2 out of the square braquets):

```suggestion
    return Drho_Dq2 (1 - upwind_factor) - upwind_factor_derivative * (current_density - upwind_density);
```